specs/dwritecore/DWriteCore.md

```diff
@@ -1050,9 +1258,591 @@ int __cdecl wmain(int argc, WCHAR** argv)
+            << indent << PropName{ "p1" } << D2D_POINT_2F{ paint.x1, paint.y1 } << '\n'
+            << indent << PropName{ "p2" } << D2D_POINT_2F{ paint.x2, paint.y2 } << '\n';
+        WriteGradient(static_cast<D2D1_EXTEND_MODE>(paint.extendMode), paint.gradientStopCount);
+        break;
+    }
+
+    case DWRITE_PAINT_TYPE_RADIAL_GRADIENT:
+    {
+        auto const& paint = element.paint.radialGradient;
+        out << indent << "DWRITE_PAINT_TYPE_RADIAL_GRADIENT:\n"
+            << indent << PropName{ "center0" } << D2D_POINT_2F{ paint.x0, paint.y0 } << '\n'
+            << indent << PropName{ "radius0" } << paint.radius0 << '\n'
+            << indent << PropName{ "center1" } << D2D_POINT_2F{ paint.x1, paint.y1 } << '\n'
+            << indent << PropName{ "radius1" } << paint.radius1 << '\n';
+        WriteGradient(static_cast<D2D1_EXTEND_MODE>(paint.extendMode), paint.gradientStopCount);
+        break;
+    }
+
+    case DWRITE_PAINT_TYPE_SWEEP_GRADIENT:
+    {
+        auto const& paint = element.paint.sweepGradient;
+        out << indent << "DWRITE_PAINT_TYPE_SWEEP_GRADIENT:\n"
+            << indent << PropName{ "center" } << D2D_POINT_2F{ paint.centerX, paint.centerY } << '\n'
+            << indent << PropName{ "startAngle" } << paint.startAngle << '\n'
+            << indent << PropName{ "endAngle" } << paint.endAngle << '\n';
+        WriteGradient(static_cast<D2D1_EXTEND_MODE>(paint.extendMode), paint.gradientStopCount);
+        break;
+    }
+
+    case DWRITE_PAINT_TYPE_GLYPH:
+    {
+        auto const& paint = element.paint.glyph;
+        out << indent << "DWRITE_PAINT_TYPE_GLYPH:\n"
+            << indent << PropName{ "glyphIndex" } << paint.glyphIndex << '\n'
+            << indent << PropName{ "child" } << '\n';
+        WriteChildren(1);
+        break;
+    }
+
+    case DWRITE_PAINT_TYPE_COLOR_GLYPH:
+    {
+        auto const& paint = element.paint.colorGlyph;
+        out << indent << "DWRITE_PAINT_TYPE_COLOR_GLYPH:\n"
+            << indent << PropName{ "glyphIndex" } << paint.glyphIndex << '\n'
+            << indent << PropName{ "clipBox" } << paint.clipBox << '\n'
+            << indent << PropName{ "child" } << '\n';
+        WriteChildren(1);
+        break;
+    }
+
+    case DWRITE_PAINT_TYPE_TRANSFORM:
+    {
+        DWRITE_MATRIX const& paint = element.paint.transform;
+        out << indent << "DWRITE_PAINT_TYPE_TRANSFORM:\n"
+            << indent << PropName{ "transform" } << paint << '\n'
+            << indent << PropName{ "child" } << '\n';
+        WriteChildren(1);
+        break;
+    }
+
+    case DWRITE_PAINT_TYPE_COMPOSITE:
+    {
+        auto const& paint = element.paint.composite;
+        out << indent << "DWRITE_PAINT_TYPE_COMPOSITE:\n"
+            << indent << PropName{ "mode" } << paint.mode << '\n'
+            << indent << PropName{ "children (source, destination)" } << '\n';
+        WriteChildren(2);
```

Comment:
We have all these implicit child counts for various paint types - this sample seems to imply that you need to know how many children are present for each type.

However, it seems that the child count will always be specified in paint.childCount for all paintTypes. Is this correct? If so, I recommend changing all instances of WriteChildren(some constant) with WriteChildren(paint.childCount) in this example. This shows a better API client practice of always using the explicit child count specified by the API rather than assuming the count for any given paint type.

If you're trying to show the expected child count for each paint type, you could perhaps mention it in a comment after each call.

Reply:
I disagree. The only paint type with a variable number of children is LAYERS. For all other paint types, the child count is known _a priori_ and is inherent to the semantics of the paint type and cannot change without breaking those semantics. For example, COMPOSITE has two inputs: the source and destination of the compositing operation. The render a COMPOSITE, a client renders the first (source) child to an intermediate, renders the second (destination) child to another intermediate, and then composes the source onto the destination using the specified mode. Given these semantics, what could you possibly do with fewer or greater than 2 children? If there were an explicit _childCount_ member, what could you do with it apart from `assert(childCount == 2)`?

Reply:
SAMPLE: Consider annotating the sample to explain some of the assumptions being made.